lib/index.js

```diff
@@ -2,115 +2,103 @@
-        return reply(Boom.badRequest('Bad HTTP authentication header format', 'Bearer'));
+        throw Boom.badRequest('Bad HTTP authentication header format', 'Bearer');
       }
 
       if (parts[0].toLowerCase() !== 'bearer') {
-        return reply(Boom.unauthorized(null, 'Bearer'));
+        throw Boom.unauthorized(null, 'Bearer');
       }
 
-      if(parts[1].split('.').length !== 3) {
-        return reply(Boom.badRequest('Bad HTTP authentication header format', 'Bearer'));
+      if (parts[1].split('.').length !== 3) {
+        throw Boom.badRequest('Bad HTTP authentication header format', 'Bearer');
       }
 
       var token = parts[1];
 
       var getKey = isFunction(settings.key) ?
         settings.key :
-        function(req, token, callback) { callback(null, settings.key); };
-
-      getKey(request, token, function(err, key, extraInfo){
-        if (err) { return reply(Boom.wrap(err)); }
-        // handle err
-        jwt.verify(token, key, settings, function(err, decoded) {
-          
-          if(err) {
-            return reply(Boom.unauthorized( 'JSON Web Token validation failed: ' + err.message, 'Bearer'));
-          }
-
-          if (!settings.validateFunc) {
-            return reply.continue({ credentials: decoded });
-          }
-
-          settings.validateFunc(decoded, extraInfo, function (err, isValid, credentials) {
+        function (req, token) { return { key: settings.key }; };
 
-            credentials = credentials || null;
+      const keyResult = await getKey(request, token);
+      const { key, extraInfo } = keyResult;
 
-            if (err) {
-              return reply(err, null, { credentials: credentials, log: { tags: ['auth', 'jwt'], data: err } });
-            }
+      let decoded;
+      try {
+        decoded = jwt.verify(token, key, settings);
+      } catch (err) {
+        throw Boom.unauthorized('JSON Web Token validation failed: ' + err.message, 'Bearer');
+      }
 
-            if (!isValid) {
-              return reply(Boom.unauthorized('Invalid token', 'Bearer'), null, { credentials: credentials });
-            }
+      if (!settings.validateFunc) {
+        return h.authenticated({ credentials: decoded });
+      }
 
-            if (!credentials || typeof credentials !== 'object') {
+      const validateResult = await settings.validateFunc(decoded, extraInfo);
 
-              return reply(Boom.badImplementation('Bad credentials object received for jwt auth validation'), null, { log: { tags: 'credentials' } });
-            }
+      const { isValid, credentials } = validateResult;
 
-            // Authenticated
+      if (!isValid) {
+        throw Boom.unauthorized('Invalid token', 'Bearer');
+      }
 
-            return reply.continue({ credentials: credentials });
-          });
+      if (!credentials || typeof credentials !== 'object') {
+        throw Boom.badImplementation('Bad credentials object received for jwt auth validation');
```

Comment:
Not sure if we need to do anything regarding the `logs`?

Reply:
It should be caught by the caller, ideally the plugin should not log

Reply:
Asked because before we passed these: `{ log: { tags: 'credentials' } }`, not sure what that is or if there's something similar on hapi 17.

Reply:
Not sure where this came from, but badImplementation has never had a 3rd argument does not has it now https://github.com/hapijs/boom#boombadimplementationmessage-data---alias-internal I've also checked the previous implementation from 1.2 and does not seem to has been there

Reply:
👍